### PR TITLE
Add several improvements to PD Forms

### PR DESCRIPTION
**Pull request by aoby**
- additional helpers
- some style improvements
- and support for dynamic required fields on both the server and client
- allow submit button text to be overridden in derived classes
- allow labels to be elements (in addition to the existing string) so they can contain richer content: formatted text, links, etc.
- only re-enable submit button in the failure case. Successful submits don't need to resubmit.
- return 409 (conflict) http status when a duplicate submit fails with [ActiveRecord::RecordNotUnique](https://app.honeybadger.io/projects/3240/faults/34602681#notice-trace) from SQL. This should also be mitigated from the above.

### CODE REVIEW

File: apps/src/code-studio/pd/form_components/FormController.jsx [hunk 1]
```diff
@@ -269,7 +268,9 @@ export default class FormController extends React.Component {
    * @returns {String[]}
    */
   getRequiredFields() {
-    return this.props.requiredFields;
+    const requiredFields = [...this.props.requiredFields];
+    this.getPageComponents().forEach(page => requiredFields.push(...page.getDynamicallyRequiredFields(this.state.data)));
```
Comment (Hamms) [inline apps/src/code-studio/pd/form_components/FormController.jsx]: why not `map` then `concat`?

File: apps/src/code-studio/pd/form_components/button_list.jsx [hunk 2]
```diff
@@ -92,15 +102,15 @@ const ButtonList = React.createClass({
         >
           {typeof answer === "object" ?
             <div>
-              <span style={{verticalAlign: 'top'}}>
-                {answerText}
+              <span style={styles.inputLabel}>
+                {answerText}{answerText.endsWith(":") ? "" : ":"}
```
Comment (Hamms) [inline apps/src/code-studio/pd/form_components/button_list.jsx]: I'm not sure I love this. Why not allow for the flexibility?

Comment (aoby) [reply apps/src/code-studio/pd/form_components/button_list.jsx]: Fair point. The form I was working on had them all end in ":", but I suppose the options can just supply that. I'll remove it.

File: apps/src/code-studio/pd/form_components/FormController.jsx (hunk 1 above)
Comment (aoby) [reply apps/src/code-studio/pd/form_components/FormController.jsx]: Because I didn't think of it :)
I like it! Will update.

**Pull request merged**